service/history/api/replication/syncworkflowstate.go

```diff
@@ -0,0 +1,293 @@
+	workflowCache wcache.Cache,
+	logger log.Logger,
+) (_ *historyservice.SyncWorkflowStateResponse, retError error) {
+	wfCtx, releaseFunc, err := workflowCache.GetOrCreateWorkflowExecution(
+		ctx,
+		shardContext,
+		namespace.ID(request.GetNamespaceId()),
+		request.Execution,
+		locks.PriorityLow,
+	)
+	defer func() {
+		if releaseFunc != nil {
+			releaseFunc(retError)
+		}
+	}()
+
+	if err != nil {
+		return nil, err
+	}
+	mu, err := wfCtx.LoadMutableState(ctx, shardContext)
+	if err != nil {
+		return nil, err
+	}
+	response := &historyservice.SyncWorkflowStateResponse{}
+	isSameBranch := workflow.TransitionHistoryStalenessCheck(mu.GetExecutionInfo().TransitionHistory, request.VersionedTransition) == nil
+	tombstoneBatch := mu.GetExecutionInfo().SubStateMachineTombstoneBatches
+	if isSameBranch &&
+		len(tombstoneBatch) > 0 && workflow.CompareVersionedTransition(tombstoneBatch[0].VersionedTransition, request.VersionedTransition) < 0 {
+		mutation, err := getMutation(mu, request.VersionedTransition)
+		if err != nil {
+			return nil, err
+		}
+		response.Attributes = &historyservice.SyncWorkflowStateResponse_Mutation{
+			Mutation: &replicationpb.SyncWorkflowStateMutationAttributes{
+				StateMutation:                     mutation,
+				ExclusiveStartVersionedTransition: request.VersionedTransition,
+			},
+		}
+	} else {
+		snapshot, err := getSnapshot(mu)
+		if err != nil {
+			return nil, err
+		}
+		response.Attributes = &historyservice.SyncWorkflowStateResponse_State{
+			State: snapshot,
+		}
+	}
+
+	newRunId := mu.GetExecutionInfo().NewExecutionRunId
+
+	releaseFunc(nil)
+	releaseFunc = nil
+
+	if len(newRunId) > 0 {
+		newRunInfo, err := getNewRunInfo(ctx, shardContext, workflowCache, namespace.ID(request.GetNamespaceId()), request.Execution.GetWorkflowId(), newRunId, logger)
+		if err != nil {
+			return nil, err
+		}
+		response.NewRunInfo = newRunInfo
+	}
+
+	events, err := getSyncStateEvents(ctx, shardContext, request.VersionHistories, mu.GetExecutionInfo().GetVersionHistories())
+	if err != nil {
+		return nil, err
+	}
+	response.EventBatches = events
+
+	return response, nil
+}
+
+func getNewRunInfo(ctx context.Context, shardContext shard.Context, workflowCache wcache.Cache, namespaceId namespace.ID, workflowId string, runId string, logger log.Logger) (_ *replicationpb.NewRunInfo, retError error) {
+	wfCtx, releaseFunc, err := workflowCache.GetOrCreateWorkflowExecution(
+		ctx,
+		shardContext,
+		namespaceId,
+		&commonpb.WorkflowExecution{
+			WorkflowId: workflowId,
+			RunId:      runId,
+		},
+		locks.PriorityLow,
+	)
+	defer func() {
+		if releaseFunc != nil {
+			releaseFunc(retError)
+		}
+	}()
+
+	if err != nil {
+		return nil, err
+	}
+	mu, err := wfCtx.LoadMutableState(ctx, shardContext)
+	switch err.(type) {
+	case nil:
+	case *serviceerror.NotFound:
+		logger.Info("new run not found", tag.WorkflowNamespaceID(namespaceId.String()), tag.WorkflowID(workflowId), tag.WorkflowRunID(runId))
```

Comment:
plz make the message is bit more descriptive. and I think we need to log the previous runID as well otherwise we have no way to investigate the issue as workflow with runID is gone.